WordPress/src/main/java/org/wordpress/android/ui/reader/discover/ReaderDiscoverFragment.kt

```diff
@@ -61,107 +60,95 @@ class ReaderDiscoverFragment : ViewPagerFragment(R.layout.reader_discover_fragme
-                is ShowReaderComments -> ReaderActivityLauncher.showReaderComments(context, it.blogId, it.postId)
-                is ShowNoSitesToReblog -> ReaderActivityLauncher.showNoSiteToReblog(activity)
-                is ShowSitePickerForResult -> ActivityLauncher
-                        .showSitePickerForResult(this@ReaderDiscoverFragment, it.preselectedSite, it.mode)
-                is OpenEditorForReblog -> ActivityLauncher
-                        .openEditorForReblog(activity, it.site, it.post, it.source)
-                is ShowBookmarkedTab -> {
-                    ActivityLauncher.viewSavedPostsListInReader(activity)
-                }
-                is ShowBookmarkedSavedOnlyLocallyDialog -> showBookmarkSavedLocallyDialog(it)
-                is ShowPostsByTag -> ReaderActivityLauncher.showReaderTagPreview(context, it.tag)
-                is ShowVideoViewer -> ReaderActivityLauncher.showReaderVideoViewer(context, it.videoUrl)
-                is ShowBlogPreview -> ReaderActivityLauncher.showReaderBlogOrFeedPreview(
-                        context,
-                        it.siteId,
-                        it.feedId
-                )
-                is ShowReportPost -> {
-                    ReaderActivityLauncher.openUrl(
-                            context,
-                            readerUtilsWrapper.getReportPostUrl(it.url),
-                            OpenUrlType.INTERNAL
-                    )
-                }
-                is ShowReaderSubs -> {
-                    ReaderActivityLauncher.showReaderSubs(context)
-                }
-                is ShowRelatedPostDetails -> Unit // Do Nothing
-            }
-        })
-        viewModel.snackbarEvents.observeEvent(viewLifecycleOwner, {
-            it.showSnackbar()
-        })
-        viewModel.preloadPostEvents.observeEvent(viewLifecycleOwner, {
-            it.addWebViewCachingFragment()
-        })
+            uiHelpers.updateVisibility(recyclerView, it.contentVisiblity)
+            uiHelpers.updateVisibility(progressBar, it.fullscreenProgressVisibility)
+            uiHelpers.updateVisibility(progressText, it.fullscreenProgressVisibility)
+            uiHelpers.updateVisibility(progressLoadingMore, it.loadMoreProgressVisibility)
+            uiHelpers.updateVisibility(actionableEmptyView, it.fullscreenEmptyVisibility)
+            ptrLayout.isEnabled = it.swipeToRefreshEnabled
+            ptrLayout.isRefreshing = it.reloadProgressVisibility
+        }
+        viewModel.navigationEvents.observeEvent(viewLifecycleOwner) { handleNavigation(it) }
+        viewModel.snackbarEvents.observeEvent(viewLifecycleOwner) { it.showSnackbar() }
+        viewModel.preloadPostEvents.observeEvent(viewLifecycleOwner) { it.addWebViewCachingFragment() }
+
         viewModel.start(parentViewModel)
     }
 
+    @Suppress("ComplexMethod")
+    private fun handleNavigation(event: ReaderNavigationEvents) = when (event) {
```

Comment:
the only thing I can think of is pulling this out into a separate class (`ReaderNavigator`). I think it's ok to suppress the warning though.